pex/common.py

```diff
@@ -15,9 +15,32 @@
 import threading
 import zipfile
 from collections import defaultdict
+from datetime import datetime
 from uuid import uuid4
 
 
+def deterministic_datetime():
+  """Return a deterministic UTC datetime object that does not depend on the system time.
+
+  First try to read from the standard $SOURCE_DATE_EPOCH, then default to midnight January 1,
```

Comment:
IMO, ever successfully reading from `SOURCE_DATE_EPOCH` would introduce a bug. So I think that it would be good to ignore any environment variables for these purposes.

Reply:
I don’t understand why this would introduce a bug? The reproducible builds project highly encourages tools like ours to let people define the value. If they don’t define it, we just use our default.

Reply:
The concern is that if I'm on a linux distribution that sets that variable, I will cache miss on a distribution that doesn't set that variable... which makes the whole thing less deterministic, rather than more. I don't see a good reason to consume it... can you share a link?

Reply:
https://reproducible-builds.org/docs/source-date-epoch/

> The concern is that if I'm on a linux distribution that sets that variable, I will cache miss on a distribution that doesn't set that variable...

Could you please elaborate on this concern? For example, in Pants it doesn't seem like this would be an issue. Our Ubuntu shards and Centos6 shards both don't set `SOURCE_DATE_EPOCH`, so I don't see any immediate concern for Pants.

More generally, this is typically a distribution-set variable rather than individual. For example, Debian sets the value as the most recent value of its changelog. So, it does not seem likely that users within an org would have an issue of cache misses.

The reason to include support for `SOURCE_DATE_EPOCH` is to be a good citizen with this standard. I don't see why a user would want to use `SOURCE_DATE_EPOCH` over our hardcoded date, but Reproducible Builds makes a big hoopla about it so it seems like a best practice to respect.

Reply:
> https://reproducible-builds.org/docs/source-date-epoch/

The motivation in these docs does not actually explain why using a hardcoded date would be less reproducible.

> Could you please elaborate on this concern? For example, in Pants it doesn't seem like this would be an issue. Our Ubuntu shards and Centos6 shards both don't set SOURCE_DATE_EPOCH, so I don't see any immediate concern for Pants.

And OSX and Windows don't. So we'd immediately have an unnecessary mismatch between our two relevant platforms.